### PR TITLE
Fix #5620 Persist format at paragraph level for new line

### LINKED ISSUE #5620: How to keep formatting between lines?
I'm trying to achieve a different behavior regarding the formatting between lines in the Lexical Editor.

### The current behavior
1. In an empty editor, the user starts adding formatting and styles in the editor (font-family, font-size, color, bold, italic, etc)
2. When you press `Enter` once, it will insert a new line, which will have an "unstyled cursor" but upon new typing, the styles and formatting are applied again to the text.
3. When you press `Enter` twice, the new line, separated by the other ones, will lose all of the previous formatting and styling, returning to the default. 

https://github.com/facebook/lexical/assets/33634787/b7fc7e60-0ce6-4d5c-be79-bea582a9a675

### The desired behavior
1. In an empty editor, the user starts adding formatting and styles in the editor (font-family, font-size, color, bold, italic, etc)
2. When pressing `Enter` regarding of how much times/lines, the styles and formats are preserved. The user doesn't have to select the font-family and font-size again (for example) every time he creates a new a paragraph.

How can I achieve this?

_Originally posted by @nickschinestzki in https://github.com/facebook/lexical/discussions/5619_

### PR DESCRIPTION
**Pull request by Sahejkm**
- The selection looses the format (bold,italic,etc) on tap of a new line
- Update the paragraph node to persist the format which could be used to update the selection format on new paragraph creation
- This fixes #5620
Before 

https://github.com/facebook/lexical/assets/163521239/389b5f78-154a-4c5d-9e2b-d523415e2789

After:

https://github.com/facebook/lexical/assets/163521239/9ccdc2ec-676d-4d02-b005-773dd8381bfd

### CODE REVIEW

File: packages/lexical/src/nodes/LexicalParagraphNode.ts [hunk 1]
```diff
@@ -6,35 +6,67 @@
...
   LexicalNode,
+  NodeKey,
 } from '../LexicalNode';
 import type {
   ElementFormatType,
   SerializedElementNode,
 } from './LexicalElementNode';
-import type {RangeSelection} from 'lexical';
 
+import {type RangeSelection} from 'lexical';
```
Comment (ivailop7) [inline packages/lexical/src/nodes/LexicalParagraphNode.ts:26]: I believe 'type' should be outside of the curly braces

File: packages/lexical/src/nodes/LexicalParagraphNode.ts [hunk 2]
```diff
@@ -6,35 +6,67 @@
...
+
+  setTextFormat(type: number): this {
+    const self = this.getWritable();
+    self.__textFormat = type;
+    return self;
+  }
+
+  hasTextFormat(type: TextFormatType): boolean {
+    const formatFlag = TEXT_TYPE_TO_FORMAT[type];
+    return (this.getTextFormat() & formatFlag) !== 0;
```
Comment (ivailop7) [inline packages/lexical/src/nodes/LexicalParagraphNode.ts:67]: Is this deliberately bitwise 'and' ?

File: packages/lexical/src/nodes/__tests__/unit/LexicalParagraphNode.test.ts [hunk 3]
```diff
@@ -113,17 +113,16 @@ describe('LexicalParagraphNode tests', () => {
...
       await editor.update(() => {
-        const result = paragraphNode.insertNewAfter(
-          $getSelection() as RangeSelection,
-          false,
-        );
-
-        expect(result).toBeInstanceOf(ParagraphNode);
-        expect(result.getDirection()).toEqual(paragraphNode.getDirection());
+        const selection = $getSelection();
+        if (selection != null && $isRangeSelection(selection)) {
```
Comment (zurfyx) [inline packages/lexical/src/nodes/__tests__/unit/LexicalParagraphNode.test.ts:117]:
```suggestion
        if ($isRangeSelection(selection)) {
```

File: packages/lexical/src/nodes/LexicalParagraphNode.ts (hunk 2 above)
Comment (zurfyx) [reply packages/lexical/src/nodes/LexicalParagraphNode.ts:72]: I believe we're mimicking the same format behavior from ElementNode, multiple values on a the same int

File: packages/lexical/src/LexicalSelection.ts [hunk 4]
```diff
@@ -837,7 +837,6 @@ export class RangeSelection implements BaseSelection {
         (firstNodeFormat !== format || firstNodeStyle !== style)
       ) {
         if (firstNode.getTextContent() === '') {
-          firstNode.setFormat(format);
```
Comment (zurfyx) [inline packages/lexical/src/LexicalSelection.ts:840]: Q: Do you know where this node format is being set now?

File: packages/lexical/src/nodes/LexicalParagraphNode.ts (hunk 2 above)
Comment (Sahejkm) [reply packages/lexical/src/nodes/LexicalParagraphNode.ts:72]: yeah its referenced from LexicalTextNode and used for logging/printing the format in the Tree View

File: packages/lexical/src/LexicalEvents.ts [hunk 5]
```diff
@@ -344,7 +344,15 @@ function onSelectionChange(
             selection.format = anchorNode.getFormat();
             selection.style = anchorNode.getStyle();
           } else if (anchor.type === 'element' && !isRootTextContentEmpty) {
-            selection.format = 0;
+            const lastNode = anchor.getNode();
+            if (
+              lastNode instanceof ParagraphNode &&
+              lastNode.getChildrenSize() === 0
+            ) {
+              selection.format = lastNode.getTextFormat();
```
Comment (zurfyx) [inline packages/lexical/src/LexicalEvents.ts:352]: We can leave this condition as children size 0 for now but we should revise it later. For example, if you start typing a paragraph with an image it wouldn't follow under this rule

**Review by zurfyx — COMMENTED** [review]
We need to look into this crash


https://github.com/facebook/lexical/assets/193447/02a1bad7-842b-4d21-a931-a10675808061

File: packages/lexical/src/LexicalSelection.ts (hunk 4 above)
Comment (Sahejkm) [reply packages/lexical/src/LexicalSelection.ts]: yeah, for text node its always from selection, same as before

File: packages/lexical-playground/__tests__/e2e/Selection.spec.mjs [hunk 6]
```diff
@@ -581,4 +582,49 @@ test.describe('Selection', () => {
       `,
     );
   });
+
+  test('Can persist the text format from the paragraph', async ({page}) => {
```
Comment (zurfyx) [inline packages/lexical-playground/__tests__/e2e/Selection.spec.mjs:586]: You probably want to skip plain text here

**Review by zurfyx — COMMENTED** [review]
LGTM, thanks! Can you add a couple more tests please?

1. Toggle format with CMD + B at the start of a paragraph of different format, toggled format persists
2. Removing all nodes of a paragraph with format and typing (I know that behavior doesn't resemble GDocs/Word but let's at least make sure it doesn't regress).

Can we follow up on the table formatting (in a separate PR)? Particularly persisting the format on creation and adding/removing rows and columns.

**Review by zurfyx — APPROVED** [review]

**Comment by nickschinestzki** [comment]
Does this fix also applies for styling? Like text color, font family, font size, etc?

**Pull request merged**